File: 1001-1100/1081-Smallest Subsequence of Distinct Characters/1081-Smallest Subsequence of Distinct Characters.cpp
class Solution {
public:
    string smallestSubsequence(string text) {
        vector<int> table(26);
        for (char c : text) {
            ++table[c - 'a'];
        }
        vector<bool> used(26);
        string result;
        for (char c : text) {
            --table[c - 'a'];
            if (used[c - 'a']) {
                continue;
            }

            while (result.size() && c < result.back() && table[result.back() - 'a']) {
                used[result.back() - 'a'] = false;
                result.pop_back();
            }

            result += c;
            used[c - 'a'] = true;
        }

        return result;
    }
};

File: 1001-1100/1081-Smallest Subsequence of Distinct Characters/1081-Smallest Subsequence of Distinct Characters.py
import collections
class Solution:
    def smallestSubsequence(self, text: str) -> str:
        table = collections.Counter(text)
        visited = set()
        result = []
        for c in text:
            table[c] -= 1
            if c in visited:
                continue
                
            while result and c < result[-1] and table[result[-1]] > 0:
                visited.remove(result.pop())
                
            result.append(c)
            visited.add(c)
            
        return ''.join(result)
